#167416898 Enable authenticated users make payments using international cards

patrickf949 opened the pull request:
## Description ##
Authenticated users should be able to comfortably make payments using international cards.
They should also be able to choose one of the payment options from international cards, nigerian cards and a tokenized card(saved card).

## Type of change ##
Please select the relevant option
- [ ] Bug fix(a non-breaking change which fixes an issue)
- [x] New feature(a non-breaking change which adds functionality)
- [ ] Breaking change(fix of a feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? ##
Please describe the tests that you ran to verify your changes
- [ ] End to end test
- [x] Integration test
- [ ] unit test

## Checklist: ##
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

## How to test this PR
- Route => `payment/international`
- Upon successful login select the `make payment` button in the nav bar and choose `international payment`
- Enter your details in the form and submit

## Related Pivotal Tracker stories ##
[#167416898](https://www.pivotaltracker.com/story/show/167416898)

patrickf949 committed:
create new component for payment and add routing

CryceTruly commented on src/app/pages/payments/payments.component.ts:
```diff
@@ -0,0 +1,15 @@
+  templateUrl: './payments.component.html',
+  styleUrls: ['./payments.component.scss']
+})
+export class PaymentsComponent implements OnInit {
```
Hey @patrickf949, There are some code climate issues here. Please Consider refactoring

patrickf949 committed:
Organise models for international payment

patrickf949 committed:
Add and configure service for international payment
Connect to service

patrickf949 committed:
Add validations for each of the fields required

patrickf949 committed:
- Add more validations
- Write tests for services and components

patrickf949 committed:
add go back functionality

patrickf949 committed:
Add more tests

Victoradukwu commented:
@patrickf949 Please rebase and resolve the merge conflicts and add a section on how to test the PR

patrickf949 committed:
Successful Rebase

patrickf949 committed:
- add links to nav bar

patrickf949 committed:
- fix failing tests after adding nav bar

patrickf949 commented:
> @patrickf949 Please rebase and resolve the merge conflicts and add a section on how to test the PR

Thanks @Victoradukwu successfully rebased

patrickf949 committed:
- Improve validation errors

patrickf949 committed:
- Fix spinner
- Add redirect to OTP page
- Fix failing tests

malep2007 commented:
@patrickf949 Update your PR description especially the testing section with details of which URL to navigate to.

malep2007 commented on src/app/pages/payments/payments.component.html:
```diff
@@ -0,0 +1,311 @@
+            (form.get('cardno').touched || form.get('cardno').dirty) &&
+            !form.get('cardno').valid
+        }" />
+      <span class="invalid-feedback">
```
This section that renders errors can be abstracted into a component of its own and use a conditional renderer to render it according to the error that exists.

patrickf949 replied:
Thanks @malep2007 done

malep2007 commented on src/app/pages/payments/payments.component.html:
```diff
@@ -0,0 +1,311 @@
+              !form.get('expirymonth').valid
+          }">
+          <option value="" value="null" disabled="true" [selected]="true">Month</option>
+          <option value="01">01</option>
```
@patrickf949 these values are hardcoded, which is unnecessary.  use the *ngFor directive and avoid repetition

patrickf949 replied:
Yeah, didn't feel right doin that.
done

malep2007 commented on src/app/pages/payments/payments.component.ts:
```diff
@@ -0,0 +1,130 @@
+    this._location.back();
+  }
+  ngOnDestroy() {
+    console.log(this.subscription[0])
```
Is it necessary to `console.log()`

patrickf949 replied:
Thanks @malep2007 done

malep2007 commented on src/app/services/internationalPayment/international-payment.service.ts:
```diff
@@ -0,0 +1,21 @@
+import { InternationalPayment } from "src/app/models";
+import { Observable } from "rxjs";
+import { HttpService } from "../http.service";
+import { HttpMethods } from 'src/app/config';
```
move all constants to the `environments` file

patrickf949 replied:
@malep2007 done

malep2007 reviewed (changes requested):
- A breakdown of the payments component is required.
- Tests need to be improved. They currently only test functionality and not what a user actually interact with.

patrickf949 committed:
- remove hardcoded month/year from form
  - implement iframe with redirect

CryceTruly commented on package.json:
```diff
@@ -56,6 +56,7 @@
     "@angular/compiler-cli": "^8.1.3",
     "@angular/language-service": "^8.1.3",
     "@babel/core": "^7.5.4",
+    "@ng-bootstrap/ng-bootstrap": "^5.1.0",
```
@patrickf949, I noticed you added this library and you did not use its components, It's also good we keep consistent with a design approach that's being used by other members.

CryceTruly commented on src/app/app.routing.ts:
```diff
@@ -7,6 +7,7 @@ import { CommonLayoutComponent } from './layouts/common-layout/common-layout.com
 import { PasswordResetComponent } from './pages/password-reset/password-reset.component';
 import { EnterResetPasswordComponent } from './pages/enter-reset-password/enter-reset-password.component';
 import { NoAuthGuard } from './guards/no-auth.guard';
+import { PaymentsComponent } from "./pages/internationalPayment/payments.component";
```
Hey @patrickf949, please run `npm run lint` to see and fix all the linting issues in your code.

CryceTruly commented on src/app/components/navbar/navbar.component.html:
```diff
@@ -48,11 +49,19 @@
     </ul>
   </div>
-</nav>
+</nav>
```
Remember to add a new line here

CryceTruly commented on src/app/models/internationalPayment.ts:
```diff
@@ -0,0 +1,14 @@
+  billingaddress: string;
+  billingstate: string;
+  billingcountry: string;
+  save_card: boolean;
```
Hey @patrickf949, please ensure to fix tslint issues.

patrickf949 replied:
Thanks @CryceTruly done

patrickf949 replied:
Great eye. @CryceTruly, however, since these values are passed to the backend I have to use the convention we chose to use for the backend

CryceTruly reviewed (changes requested):
@patrickf949, please remember to add an appropriate title to your page, you can use the title service to achieve that.

patrickf949 committed:
- Reorganize work to prevent conflicts

Victoradukwu commented on src/app/app.routing.ts:
```diff
@@ -8,6 +8,7 @@ import { CommonLayoutComponent } from './layouts/common-layout/common-layout.com
 import { PasswordResetComponent } from './pages/password-reset/password-reset.component';
 import { EnterResetPasswordComponent } from './pages/enter-reset-password/enter-reset-password.component';
 import { NoAuthGuard } from './guards/no-auth.guard';
+import { PaymentsComponent } from "./pages/payment/internationalPayment.component";
```
@patrickf949 As agreed during my sync with you, let's have a `payment.routing.ts` and a `payment.module.ts` files in the **payment** folder and register all payment-related routes and components there.

Victoradukwu commented on src/app/app.module.ts:
```diff
@@ -18,6 +18,9 @@ import { EnterResetPasswordComponent } from './pages/enter-reset-password/enter-
 import { AuthGuard } from './guards/auth.guard';
 import { NoAuthGuard } from './guards/no-auth.guard';
 
+
```
@patrickf949 It does not look good to have blank lines in between import statements

Victoradukwu commented on src/app/pages/enter-reset-password/enter-reset-password.component.spec.ts:
```diff
@@ -5,7 +5,7 @@ import { PasswordResetComponent } from '../password-reset/password-reset.compone
 import { FormsModule, ReactiveFormsModule } from '@angular/forms';
 import { HttpClientModule } from '@angular/common/http';
-import { DebugElement } from '@angular/core';
+import { DebugElement, CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA } from '@angular/core';
```
@patrickf949 Remove the **NO_ERRORS_SCHEMA** as you are not using it anywhere.

Victoradukwu commented on src/app/pages/enter-reset-password/enter-reset-password.component.spec.ts:
```diff
@@ -15,6 +15,8 @@ import { AuthLayoutComponent } from 'src/app/layouts/auth-layout/auth-layout.com
 import { CommonLayoutComponent } from 'src/app/layouts/common-layout/common-layout.component';
 import { ComponentsModule } from 'src/app/components/components.module';
 import { ToastrService } from 'ngx-toastr';
+import { PaymentsComponent } from '../payment/internationalPayment.component';
```
@patrickf949 change this to absolute import

Victoradukwu commented on src/app/pages/password-reset/password-reset.component.spec.ts:
```diff
@@ -15,6 +15,8 @@ import { ComponentsModule } from 'src/app/components/components.module';
 import { DebugElement } from '@angular/core';
 import { throwError, of } from 'rxjs';
 import { ToastrService } from 'ngx-toastr';
+import { PaymentsComponent } from '../payment/internationalPayment.component';
```
Use absolute import here.

Victoradukwu commented on src/app/pages/payment/internationalPayment.component.ts:
```diff
@@ -0,0 +1,152 @@
+  templateUrl: "./internationalPayment.component.html",
+  styleUrls: ["./internationalPayment.component.scss"]
+})
+export class PaymentsComponent implements OnInit {
```
Since this component handles only international payment, it'd be better to name it appropriately so. maybe `InternationalPayment`. The file name should also be hyphenated, `international-payment.component.ts`. Likewise the associated style and template classes.

Victoradukwu commented on src/app/pages/payment/internationalPayment.component.ts:
```diff
@@ -0,0 +1,152 @@
+
+
+  onSubmitPaymentDetails({ value }) {
+    console.log(value);
```
Why are we console-logging **value**?

patrickf949 replied:
Thanks @Victoradukwu done

Victoradukwu commented on src/app/pages/payment/internationalPayment.component.ts:
```diff
@@ -0,0 +1,152 @@
+  constructor(
+    private internationalPaymentService: InternationalPaymentService,
+    private toastrService: ToastrService,
+    private _location: Location,
```
Any reason we are preceding the variable **location** with an underscore?

patrickf949 replied:
Thanks @Victoradukwu done

Victoradukwu commented on src/app/services/internationalPayment/international-payment.service.spec.ts:
```diff
@@ -0,0 +1,68 @@
+  HttpTestingController
+} from "@angular/common/http/testing";
+import { ReactiveFormsModule, FormGroup, FormsModule } from "@angular/forms";
+import { environment } from "../../../environments/environment";
```
Use absolute import here

patrickf949 replied:
Thanks @Victoradukwu done

Victoradukwu reviewed (changes requested):
@patrickf949 Nice job. Please work on the requested changes.

Tevinthuku commented on src/app/pages/payment/internationalPayment.component.ts:
```diff
@@ -0,0 +1,153 @@
+
+  generateYears() {
+    const thisYear = new Date().getFullYear();
+    this.expectedYears = Array.apply(0, Array(14))
```
Hey @patrickf949 ,great job working on the feature, I may have an alternative for this method. We can use `Array.from(Array(14).keys())` because it generates an array of numbers and as a result, we do not need to have `_element` and `index` properties in the map function, we could just have 1 property. This can also work in your favor since tslint is compaining that `_element` should be camelCased
Here's a screenshot demonstrating the functionality that I'm proposing and how you can use it 

![image](https://user-images.githubusercontent.com/12128153/63406996-20b9e100-c3f4-11e9-8b95-cf6aebd48fad.png)

patrickf949 replied:
> Hey @patrickf949 ,great job working on the feature, I may have an alternative for this method. We can use `Array.from(Array(14).keys())` because it generates an array of numbers and as a result, we do not need to have `_element` and `index` properties in the map function, we could just have 1 property. This can also work in your favor since tslint is compaining that `_element` should be camelCased
> Here's a screenshot demonstrating the functionality that I'm proposing and how you can use it
> 
> ![image](https://user-images.githubusercontent.com/12128153/63406996-20b9e100-c3f4-11e9-8b95-cf6aebd48fad.png)

Thanks @Tevinthuku done

patrickf949 commented:
> @patrickf949 Nice job. Please work on the requested changes.

Thanks @Victoradukwu done

alexxsanya commented on src/app/components/navbar/navbar.component.scss:
```diff
@@ -174,4 +174,37 @@ ul.top-menu > li > a:hover .caret {
+  z-index: 1;
+}
+
+.dropdown-content a {
```
@patrickf949 since we are using SCSS, I recommend you consider wrapping styles of children elements under there parent class name. i.e for `dropdown-content` you define the children classes  like below
```
.dropdown-content {
  a {
       color: black;
       &: hover {

       }
    }
}
```

malep2007 commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,150 @@
+
+    this.generateYears();
+    this.form = this.fb.group({
+      cardno: [
```
Variable names are not done well, consider refactoring, use `camelCase`

patrickf949 replied:
thanks @malep2007 done

malep2007 commented:
General Comments
- The routing of the URLs is not consistent, change from `international-payment` to something like `/payment/international`
- **Pet peeve:** Forgot to add new line to the  `navbar.component.html` file.
- Poor variable naming conventions, Stick to camel case for example camelCase  should be used.

patrickf949 committed:
- add reuseable component for handling input errors
- refactor generation of years of expiry of clients' cards
- fix failing tests
- use camelCasing for all variables other than those sent to the backend

patrickf949 commented:
> General Comments
> 
> * The routing of the URLs is not consistent, change from `international-payment` to something like `/payment/international`
> * **Pet peeve:** Forgot to add new line to the  `navbar.component.html` file.
> * Poor variable naming conventions, Stick to camel case for example camelCase  should be used.

Thanks @malep2007 done

patrickf949 committed:
- add status of transaction upon successful completion

patrickf949 committed:
- fix tests

patrickf949 committed:
- styling

Tevinthuku reviewed (commented):
Hi @patrickf949  , Im not so sure this drop down on top of another dropdown would be awesome

![image](https://user-images.githubusercontent.com/12128153/63520658-c147f800-c4fd-11e9-906e-b3d2bf54fe86.png)



Its weird coz it disappears and reappears whenever I hover my Name

patrickf949 commented:
> Hi @patrickf949 , Im not so sure this drop down on top of another dropdown would be awesome
> 
> ![image](https://user-images.githubusercontent.com/12128153/63520658-c147f800-c4fd-11e9-906e-b3d2bf54fe86.png)
> 
> Its weird coz it disappears and reappears whenever I hover my Name

Hello @Tevinthuku this was already changed.
Please hard reload that page
or use this for the deployed application
[here](https://landville-frontend-pr-19.herokuapp.com/payment/international)

Tevinthuku commented:
> > Hi @patrickf949 , Im not so sure this drop down on top of another dropdown would be awesome
> > ![image](https://user-images.githubusercontent.com/12128153/63520658-c147f800-c4fd-11e9-906e-b3d2bf54fe86.png)
> > Its weird coz it disappears and reappears whenever I hover my Name
> 
> Hello @Tevinthuku this was already changed.
> Please hard reload that page
> or use this for the deployed application
> [here](https://landville-frontend-pr-19.herokuapp.com/payment/international)

Hey sorry about that, I just realized I got the outdated version and the UI had changed

Tevinthuku reviewed (approved):
This works as expected. Nice work @patrickf949

Victoradukwu commented on src/app/app.module.ts:
```diff
@@ -19,7 +19,8 @@ import { PasswordResetComponent } from './pages/password-reset/password-reset.co
-
+import { InternationalPaymentStatusComponent } from './pages/payment/international-payment-status/international-payment-status.component';
+import { PaymentModule } from './pages/payment/payment.module';
 
```
**PaymentModule** is already imported and registered with the **CommonLayoutModule**. We should not have it here anymore.
Also why are not declaring **InternationalPaymentStatusComponent** in the `PaymentModule` (lines 23 and 46)?

patrickf949 replied:
*InternationalPaymentStatusComponent* is in the app module because for that particular component I did not want to have the header and footer as they would appear twice on the particular page where the link is called

Victoradukwu commented on src/app/app.module.ts:
```diff
@@ -31,6 +32,7 @@ import { NoAuthGuard } from './guards/no-auth.guard';
     RouterModule,
     ComponentsModule,
     HttpClientModule,
+    NgxSpinnerModule,
```
Duplicate. This is already present on line 38

patrickf949 replied:
thanks @Victoradukwu done

Victoradukwu commented on src/app/app.routing.ts:
```diff
@@ -47,6 +48,7 @@ const routes: Routes = [
     path: 'auth/password-reset',
     component: EnterResetPasswordComponent
   },
+  { path: 'international-payment/status', component: InternationalPaymentStatusComponent},
```
Let's keep all payment-related routes in `payment.routing.ts`

patrickf949 replied:
This is also because I did not want to have a header and footer on that particular page

Victoradukwu commented on src/app/layouts/auth-layout/auth-layout.module.ts:
```diff
@@ -59,7 +59,8 @@ export function provideConfig() {
     RegisterFormComponent,
     RegisterHeaderComponent,
-    SocialLoginComponentt
+    SocialLoginComponentt,
```
I believe we do not need the **comma** and the blank line.

Victoradukwu commented on src/app/pages/home/home.component.scss:
```diff
@@ -185,4 +185,4 @@ hr.br-line {
 		}
 	}
-}
+}
```
Let's have a blank line here

Victoradukwu commented on src/app/pages/payment/international-payment-status/international-payment-status.component.scss:
```diff
@@ -0,0 +1,3 @@
+.text-failure{
+    color:red;
+}
```
Add a blank line

Victoradukwu commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,151 @@
+      billingState: [null, [Validators.required]],
+      billingCountry: [null, [Validators.required]],
+      saveCard: false,
+      purpose: [null, [Validators.required]]
```
We should also have a `non-required` field **properry_id**. This will become required if if the chosen purpose is **Buying**

patrickf949 replied:
Does this also mean I activate the buying feature?

Victoradukwu commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,151 @@
+  FormBuilder,
+  Validators,
+  ReactiveFormsModule,
+  AbstractControl
```
ReactiveFormsModule and AbstractControl are not used anywhere here. We should not import them

Victoradukwu commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,151 @@
+  AbstractControl
+} from '@angular/forms';
+import { NgxSpinnerService } from 'ngx-spinner';
+import { Router } from '@angular/router';
```
Unused import **Router**

patrickf949 committed:
- more styling
- rebased

Victoradukwu commented on src/app/services/SocialAuth/socialauth.service.spec.ts:
```diff
@@ -1,15 +1,15 @@
+} from "@angular/common/http/testing";
+import { environment } from "src/environments/environment";
+import { of, Observable } from "rxjs";
+import { APPCONFIG } from "src/app/config";
```
@patrickf949 Why are we changing single quotes to double quotes?

Victoradukwu commented on src/app/services/payment/international-payment.service.ts:
```diff
@@ -0,0 +1,19 @@
+      payment
+    );
+  }
+}
```
@patrickf949 This should be in the existing `payment-service` file. Just add **createInternationalPayment** method to the **PaymentService** class.

patrickf949 replied:
thanks @Victoradukwu Done

Victoradukwu commented on src/app/shared/input-errors/input-errors.component.html:
```diff
@@ -0,0 +1,6 @@
+    {{ error }}
+  </span>
+  <br />
+</span>
```
Put a blank line here

Victoradukwu commented on src/app/shared/input-errors/input-errors.component.scss:
```diff
@@ -0,0 +1,6 @@
+    margin-top: .25rem;
+    font-size: 80%;
+    color: #dc3545;
+}
```
Add a blank line here

Victoradukwu reviewed (changes requested):
Great improvement, @patrickf949. Take a look at the reviews I left.

CryceTruly commented on src/app/app.module.ts:
```diff
@@ -18,29 +18,32 @@ import { PasswordResetComponent } from './pages/password-reset/password-reset.co
 import { AuthGuard } from './guards/auth.guard';
 import { NoAuthGuard } from './guards/no-auth.guard';
-
+import { InternationalPaymentStatusComponent } from './pages/payment/international-payment-status/international-payment-status.component';
```
Hey @patrickf949, we should use an absolute import here

patrickf949 replied:
Thanks @CryceTruly done

CryceTruly commented on src/app/components/navbar/navbar.component.html:
```diff
@@ -41,30 +79,59 @@
+        >
           <span class="profile-ava">
-            <img [src]="profileImage" alt="avatar image" class="rounded-circle z-depth-0" height="35">
+            <img
```
Hey @patrickf949, we need to add a descriptive alt text to the images for when they cant display.

patrickf949 replied:
Yes @CryceTruly , thats great, please suggest what may be more convenient since there are actually multiple use cases of images in the application, though for this feature currently I haven't worked with any

CryceTruly commented on src/app/pages/payment/international-payment-status/international-payment-status.component.spec.ts:
```diff
@@ -0,0 +1,31 @@
+      imports: [RouterTestingModule],
+      declarations: [ InternationalPaymentStatusComponent ],
+      providers: []
+      //   { provide: ActivatedRoute, useValue: activatedRouteSpy }, ]
```
@patrickf949, let's remove the commented code

patrickf949 replied:
@CryceTruly done

CryceTruly commented on src/app/services/SocialAuth/socialauth.service.spec.ts:
```diff
@@ -1,15 +1,15 @@
+import { APPCONFIG } from "src/app/config";
 
-describe('Test for Loginservice', () => {
+describe("Test for Loginservice", () => {
```
Hey @patrickf949, let's use conventional ts single quotes hear.

patrickf949 replied:
@CryceTruly done

CryceTruly commented on src/app/services/SocialAuth/socialauth.service.spec.ts:
```diff
@@ -1,15 +1,15 @@
+} from "@angular/common/http/testing";
+import { environment } from "src/environments/environment";
+import { of, Observable } from "rxjs";
+import { APPCONFIG } from "src/app/config";
```
@patrickf949 , let us move everything in the config file into the environments file, delete the file and refactor all usages accordingly.

patrickf949 replied:
Thanks @CryceTruly done

patrickf949 committed:
- fix lintin errors as per feedback
- removed unused imports
- add field of property id

patrickf949 committed:
- move all payment services to one file
- style status page

Victoradukwu commented on src/app/helpers/spies.ts:
```diff
@@ -64,3 +64,14 @@ export const propertiesServiceSpy = createSpyObj('PropertiesService', [
 
 export const httpHandlerSpy = jasmine.createSpyObj('HttpHandler', ['handle']);
+
+export const internationalPaymentServiceSpy = createSpyObj(
```
@patric I believe we no loger need this

Victoradukwu commented on src/app/layouts/common-layout/common-layout.routing.ts:
```diff
@@ -6,6 +6,7 @@ import { NoPropertiesComponent } from 'src/app/pages/properties/no-properties/no
 import { RegistersuccessComponent } from 'src/app/pages/registration/registersuccess/registersuccess.component';
 import { PropertiesComponent } from '../../pages/properties/properties.component';
 import { AuthGuard } from 'src/app/guards/auth.guard';
+import { InternationalPaymentComponent } from 'src/app/pages/payment/international-payment/international-payment.component';
```
@patrickf949 Register this in the Payment routing file and remove it from here

Victoradukwu commented on src/app/services/payment-service.spec.ts:
```diff
@@ -1,4 +1,4 @@
-import { PaymentService } from './payment-service';
+import { PaymentService } from './payment/payment-service';
```
@patrickf949 If you are moving the service file to a different folder, ensure to move it's spec file there too and remove it from here.

Victoradukwu commented on src/app/components/navbar/navbar.component.html:
```diff
@@ -41,30 +79,60 @@
+            routerLinkActive="active"
+          >
+            <i class="fa fa-sign-out"></i> Log Out</a
+          >
```
@patrickf949  Check the navbar carefully, you have overridden an existing styling.

patrickf949 replied:
Thanks done.

Victoradukwu reviewed (changes requested):
@patrickf949 Please take a look at the issues raised.

Tevinthuku commented on src/environments/environment.ts:
```diff
@@ -6,5 +6,6 @@ export const environment = {
     'https://landville-backend-web-api.herokuapp.com/api/v1/auth/profile/',
-  api_url: 'https://landville-backend-web-api.herokuapp.com/api/v1',
+  // api_url: 'https://landville-backend-web-api.herokuapp.com/api/v1',
+  api_url: 'http://localhost:8000/api/v1',
```
You might have left this here for testing purposes. If this gets merged, things will break

patrickf949 replied:
for now while heroku is down I'm using localhost

Tevinthuku replied:
👍 , just don't forget it there

Tevinthuku commented on src/app/pages/payment/international-payment-status/international-payment-status.component.ts:
```diff
@@ -0,0 +1,24 @@
+  styleUrls: ['./international-payment-status.component.scss']
+})
+export class InternationalPaymentStatusComponent implements OnInit {
+  imgSrc: any;
```
`imgSrc` should be of type `string`. Type `any` can allow someone, later on, to change the value to something other than a string and typescript would still allow it, but it would end up breaking your implementation

Tevinthuku commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,152 @@
+  form: FormGroup;
+  formData;
+  subscription: Subscription[] = [];
+  currentUrl: SafeUrl;
```
you aren't using `currentUrl`, do we need it?

patrickf949 replied:
@Tevinthuku done

Tevinthuku commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,152 @@
+  loading: boolean;
+  payload: any;
+  form: FormGroup;
+  formData;
```
`formData` is unused here. You can get rid of it if you arent using it

patrickf949 replied:
@Tevinthuku done

Tevinthuku commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,152 @@
+            this.loading = false;
+
+            let errors = '';
+            for (const key in error) {
```
`Ts-lint ` is complaining here.
![image](https://user-images.githubusercontent.com/12128153/63682688-269f3000-c801-11e9-8dfd-6a60bc7dea70.png)

What do you think of using `object.keys` or `object.entries` here. so that you can iterate through the error object with `.map`

https://appdividend.com/2019/03/15/javascript-object-values-example-object-values-tutorial/

patrickf949 replied:
All done @Tevinthuku

Tevinthuku commented on src/app/shared/input-errors/input-errors.component.html:
```diff
@@ -0,0 +1,6 @@
+<span class="custom-error">
```
Hey, @patrickf949 , this component isn't doing much here, 
what do you think of having a `p` tag on the HTML template you are using this and conditionally render it incase of an error

patrickf949 replied:
Yes @Tevinthuku, that makes sense but, this is simply a reusable component that anyone can make use of without having to style

patrickf949 committed:
- redirect user after validation

patrickf949 committed:
- style iframe for responsiveness

Tevinthuku commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,155 @@
+    private fb: FormBuilder,
+    private spinner: NgxSpinnerService,
+    private sanitizer: DomSanitizer,
+    private router: Router,
```
`Router` is not used

patrickf949 replied:
@Tevinthuku done

Tevinthuku commented on src/app/pages/payment/international-payment/international-payment.component.ts:
```diff
@@ -0,0 +1,152 @@
+      purpose: value.purpose,
+    };
+
+    value.propertyId ? this.payload.property_id = this.propertyId : null;
```
Im not so sure what this expression is doing, but if you are setting the `this.payload.property_id` to `this.propertyId` if `value.propertyId` is not null then here is a better way of doing so

```js
    this.payload = {
      cardno: value.cardNo,
      cvv: value.cvv,
      expirymonth: value.expiryMonth,
      expiryyear: value.expiryYear.slice(-2),
      amount: value.amount,
      billingzip: value.billingZip,
      billingcity: value.billingCity,
      billingaddress: value.billingAddress,
      billingstate: value.billingState,
      billingcountry: value.billingCountry,
      save_card: value.saveCard,
      purpose: value.purpose,
      property_id: value.propertyId && this.propertyId
    };
```

its much cleaner and also gets rid of the following `ts-lint` issue

![image](https://user-images.githubusercontent.com/12128153/63788323-f38e9680-c8fd-11e9-9c74-31d63f9ebf67.png)

Victoradukwu reviewed (approved):
LGTM

alexxsanya reviewed (changes requested):
<img width="1007" alt="Screenshot 2019-08-27 at 18 09 57" src="https://user-images.githubusercontent.com/21230051/63783571-e7064000-c8f5-11e9-8b75-c2d985a4b869.png">
@patrickf949 I notice the review app is failing. kindly check it out.

patrickf949 commented:
> <img alt="Screenshot 2019-08-27 at 18 09 57" width="1007" src="https://user-images.githubusercontent.com/21230051/63783571-e7064000-c8f5-11e9-8b75-c2d985a4b869.png">
> 
> @patrickf949 I notice the review app is failing. kindly check it out.

@alexxsanya yes that is because only one dyno was paid for on heroku which can only be used by develop branch.
You can test the app locally but change the `api_url` to `http:localhost:8000/api/v1` in the `environment.ts` file

alexxsanya reviewed (changes requested):
<img width="735" alt="Screenshot 2019-08-28 at 10 45 57" src="https://user-images.githubusercontent.com/21230051/63835964-43627180-c981-11e9-9366-cd12fb3828b9.png">
@patrickf949 well done on this. I like the form validation though I have some suggestions : 

- for the fields which require numbers/integers, change the input field to `number`.
- The input fields look squeezed to each, some spacing between themselves I believe would make it look better.

CryceTruly reviewed (approved):
Hey @patrickf949, great work here. LGTM
